File: frontend/services/portfolio/portfolioService.ts
import { CreateOut, UpdateOut, DeleteOut } from "@/schemas/common/types";
import {
  PortfolioCreateInSchema,
  PortfolioUpdateInSchema,
  PortfoliosOut,
  PortfolioOut,
  HoldingCreateInSchema,
  HoldingUpdateInSchema,
  HoldingsOut,
} from "@/schemas/portfolio/portfolio";
import { apiCall } from "@/utils/common/api/client";
import { handleZodValidationError, validateWithZod } from "@/lib/zod/validation";

const BASE_URL = "/api/external/backend/portfolio";

const stringifyGridParams = (params: any): Record<string, any> => {
  const queryParams: Record<string, any> = { ...params };
  if (queryParams.filter) queryParams.filter = JSON.stringify(queryParams.filter);
  if (queryParams.sort) queryParams.sort = JSON.stringify(queryParams.sort);
  return queryParams;
};

interface SelectListOptions {
  /** #332 — 서버 실패(`success:false`)와 정상 응답 0건을 구분해야 하는 호출부용. 기본 false. */
  throwOnFailure?: boolean;
}

// ── Portfolio (master) ─────────────────────────────────────────────────
export const selectPortfolioList = async (params: any, options?: SelectListOptions): Promise<PortfoliosOut | null> => {
  return apiCall<PortfoliosOut>(BASE_URL, {
    method: "GET",
    params: stringifyGridParams(params),
    throwOnFailure: options?.throwOnFailure,
  });
};

export const selectPortfolio = async (data: any): Promise<PortfolioOut | null> => {
  return apiCall<PortfolioOut>(`${BASE_URL}/${data.portfolio_id}`, { method: "GET" });
};

export const createPortfolio = async (data: any): Promise<CreateOut | null> => {
  try {
    const validatedData = validateWithZod(PortfolioCreateInSchema, data);
    return apiCall<CreateOut>(BASE_URL, { method: "POST", data: validatedData });
  } catch (error) {
    handleZodValidationError(error);
  }
};

export const updatePortfolio = async (data: any): Promise<UpdateOut | null> => {
  try {
    const { portfolio_id, ...baseData } = data;
    const validatedData = validateWithZod(PortfolioUpdateInSchema, baseData);
    return apiCall<UpdateOut>(`${BASE_URL}/${portfolio_id}`, { method: "PUT", data: validatedData });
  } catch (error) {
    handleZodValidationError(error);
  }
};

export const deletePortfolio = async (data: any): Promise<DeleteOut | null> => {
  return apiCall<DeleteOut>(`${BASE_URL}/${data.portfolio_id}`, { method: "DELETE" });
};

// ── Holding (detail) ───────────────────────────────────────────────────
export const selectHoldingList = async (params: any, options?: SelectListOptions): Promise<HoldingsOut | null> => {
  const { portfolio_id, ...rest } = params;
  return apiCall<HoldingsOut>(`${BASE_URL}/${portfolio_id}/holding`, {
    method: "GET",
    params: stringifyGridParams(rest),
    throwOnFailure: options?.throwOnFailure,
  });
};

export const createHolding = async (data: any): Promise<CreateOut | null> => {
  try {
    const { portfolio_id, ...baseData } = data;
    const validatedData = validateWithZod(HoldingCreateInSchema, baseData);
    return apiCall<CreateOut>(`${BASE_URL}/${portfolio_id}/holding`, { method: "POST", data: validatedData });
  } catch (error) {
    handleZodValidationError(error);
  }
};

export const updateHolding = async (data: any): Promise<UpdateOut | null> => {
  try {
    const { portfolio_id, ticker, ...baseData } = data;
    const validatedData = validateWithZod(HoldingUpdateInSchema, baseData);
    return apiCall<UpdateOut>(`${BASE_URL}/${portfolio_id}/holding/${ticker}`, {
      method: "PUT",
      data: validatedData,
    });
  } catch (error) {
    handleZodValidationError(error);
  }
};

export const deleteHolding = async (data: any): Promise<DeleteOut | null> => {
  return apiCall<DeleteOut>(`${BASE_URL}/${data.portfolio_id}/holding/${data.ticker}`, { method: "DELETE" });
};

File: backend-service/app/services/portfolio/portfolio_service.py
from core.auth_context import require_workspace_id
from core.exceptions import ConflictError, NotFoundError
from repositories.portfolio.portfolio_repository import PortfolioRepository


class PortfolioService:
    def __init__(self, portfolio_repository: PortfolioRepository):
        self.portfolio_repository = portfolio_repository

    # ── Portfolio (master) ─────────────────────────────────────────────
    def select_portfolio_list(self, args: dict) -> tuple[list, int]:
        args["workspace_id"] = require_workspace_id()
        return self.portfolio_repository.select_portfolio_list(args)

    def select_portfolio(self, args: dict) -> dict:
        args["workspace_id"] = require_workspace_id()
        portfolio = self.portfolio_repository.select_portfolio(args)
        if not portfolio:
            raise NotFoundError("데이터를 찾을 수 없습니다.")
        return portfolio

    def insert_portfolio(self, args: dict) -> tuple:
        args["workspace_id"] = require_workspace_id()
        if self.portfolio_repository.select_portfolio(args):
            raise ConflictError("이미 존재하는 데이터입니다.")
        return self.portfolio_repository.insert_portfolio(args)

    def update_portfolio(self, args: dict) -> None:
        args["workspace_id"] = require_workspace_id()
        if not self.portfolio_repository.select_portfolio(args):
            raise NotFoundError("데이터를 찾을 수 없습니다.")
        self.portfolio_repository.update_portfolio(args)

    def delete_portfolio(self, args: dict) -> None:
        args["workspace_id"] = require_workspace_id()
        if not self.portfolio_repository.select_portfolio(args):
            raise NotFoundError("데이터를 찾을 수 없습니다.")
        self.portfolio_repository.delete_portfolio(args)

    # ── Holding (detail) ───────────────────────────────────────────────
    def select_holding_list(self, args: dict) -> tuple[list, int]:
        args["workspace_id"] = require_workspace_id()
        return self.portfolio_repository.select_holding_list(args)

    def select_holding(self, args: dict) -> dict:
        args["workspace_id"] = require_workspace_id()
        holding = self.portfolio_repository.select_holding(args)
        if not holding:
            raise NotFoundError("데이터를 찾을 수 없습니다.")
        return holding

    def insert_holding(self, args: dict) -> tuple:
        args["workspace_id"] = require_workspace_id()
        # 저장소에 FK 가 없다 — 여기서 안 막으면 부모 없는 보유종목이 남고 API 로 지울 수도 없다.
        if not self.portfolio_repository.select_portfolio(args):
            raise NotFoundError("포트폴리오를 찾을 수 없습니다.")
        if self.portfolio_repository.select_holding(args):
            raise ConflictError("이미 존재하는 데이터입니다.")
        return self.portfolio_repository.insert_holding(args)

    def update_holding(self, args: dict) -> None:
        args["workspace_id"] = require_workspace_id()
        if not self.portfolio_repository.select_holding(args):
            raise NotFoundError("데이터를 찾을 수 없습니다.")
        self.portfolio_repository.update_holding(args)

    def delete_holding(self, args: dict) -> None:
        args["workspace_id"] = require_workspace_id()
        if not self.portfolio_repository.select_holding(args):
            raise NotFoundError("데이터를 찾을 수 없습니다.")
        self.portfolio_repository.delete_holding(args)
